roachtest: default Run{,E} to a per-invocation file

tbg opened the pull request:
A regular complaint about roachtest is that it is difficult to reason about
the flow of the test given only the log files. This is in part because the
main log often contains lenghty workload outputs that create large files
with lots of non-timestamped sections where output is often even hard to
assign to a specific process. Questions that often come up are "did this
process already terminate at this point?", "Did it ever print output?",
"What process does this log line belong to", "What is this test even doing?
All I'm seeing are pages and pages of some bogus workload output",among
many others.

This commit changes the default behavior to creating a unique file per
invocation (sorting in timestamp order) and streaming the output there and
only there. Where streaming to stdout/stderr is desired, `RunL` can be used
instead.

Certainly a lot more can be done here, but this could already help quite a
bit in keeping the main log manageable in size while also allowing us to
better diagnose test failures.

Release note: None

tbg committed:
roachtest: default Run{,E} to a per-invocation file

A regular complaint about roachtest is that it is difficult to reason
about the flow of the test given only the log files. This is in part
because the main log often contains lenghty workload outputs that create
large files with lots of non-timestamped sections where output is often
even hard to assign to a specific process. Questions that often come up
are "did this process already terminate at this point?", "Did it ever
print output?", "What process does this log line belong to", "What is
this test even doing? All I'm seeing are pages and pages of some bogus
workload output",among many others.

This commit changes the default behavior to creating a unique file per
invocation (sorting in timestamp order) and streaming the output there
and only there. Where streaming to stdout/stderr is desired, `RunL` can
be used instead.

Certainly a lot more can be done here, but this could already help quite
a bit in keeping the main log manageable in size while also allowing us
to better diagnose test failures.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43129)
<!-- Reviewable:end -->

tbg commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -1806,7 +1806,25 @@ func (c *cluster) Install(
+	// NB: we set no prefix because it's only going to a file anyway.
+	l, err := c.l.ChildLogger(logFile, quietStderr, quietStdout)
+	if err != nil {
+		return c.RunL(ctx, c.l, node, args...)
```
I should probably just return the error here. Is it ever expected that ChildLogger fails?

andreimatei reviewed (commented):
For the 2nd commit, I think it's great to cut down the number of logger instantiations, but I like passing the  loggers explicitly to various methods....

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-:-LvvCIRR8ftkOfRpwU4m:b-16l8yi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/cmd/roachtest/cluster.go, line 1807 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvvBImj43TdSdmdM0FT:-LvvBImk59GDP3oMasvH:b-roz6wz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a1451c06665452c8f1334ff2151f94a655f6af/pkg/cmd/roachtest/cluster.go#L1807)):*
> ```Go
> }
> 
> // RunE runs a command on the specified node, returning an error.
> ```

please add a comment about where the log goes
___
*[pkg/cmd/roachtest/cluster.go, line 1811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvvBajLEAPgvl9duAfo:-LvvBajLEAPgvl9duAfp:bl33amy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a1451c06665452c8f1334ff2151f94a655f6af/pkg/cmd/roachtest/cluster.go#L1811)):*
> ```Go
> 	dest := fmt.Sprintf(`run_%s_%s_%s`,
> 		timeutil.Now().Format(`15:04:05.000`),
> 		c.makeNodes(node),
> ```

what's the point of this `makeNodes()` call?
___
*[pkg/cmd/roachtest/cluster.go, line 1818 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvuNlh4QSl5gVOktuk3-r2-1818:-LvvBrzTAg-_kCFcvATO:b-ejt591) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d43236c1f53443e22b9bc122c5535d1c3df85c/pkg/cmd/roachtest/cluster.go#L1818)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I should probably just return the error here. Is it ever expected that ChildLogger fails?
</blockquote></details>

return err indeed


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
In fact, I hope to get rid of `c.l`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-:-LvvCVKr1znz8AbJ5hUt:bh7gcnh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR

For the record, Andrei and I chatted offline and I got an LGTM. We are interested in cleaning up the interplay between `c.l`, `c.test`, `test.l` to make sure we get rid of all of this confusion, but this is out of scope here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-:-Lyctaae0Xhql4vjUwAr:bs4z06r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/cmd/roachtest/cluster.go, line 1807 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvvBImj43TdSdmdM0FT:-Lyct_pKDmKdS1pzIlOf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a1451c06665452c8f1334ff2151f94a655f6af/pkg/cmd/roachtest/cluster.go#L1807)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment about where the log goes
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/cluster.go, line 1811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvvBajLEAPgvl9duAfo:-Lycsnrc6IxeQVFz6gD8:b-en831e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a1451c06665452c8f1334ff2151f94a655f6af/pkg/cmd/roachtest/cluster.go#L1811)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what's the point of this `makeNodes()` call?
</blockquote></details>

It makes the filename describe the full command. Without this, the filename wouldn't reflect where the command was run. For example

`c.Run(ctx, c.Range(1,3), "echo", "foo")` would create some file `15_04_05.000_echo_foo`, but the actual code will make it `15_04_05.000_1-3_echo_foo`. I think this will be helpful sometimes (and it doesn't hurt if it's not).


<!-- Sent from Reviewable.io -->

tbg committed:
issues: don't try to post of GH token is set but empty

We sometimes pass an empty token in because we use

``
env GITHUB_TOKEN="${GITHUB_TOKEN-}" ./build/builder.sh ...
```

and it's easier to improve the check than to try to wrangle the scripts
to only declare the env var if it's nonempty.

Release note: None

tbg commented:
bors r=andreimatei

Getting very interested in merging this again while looking at all of the tests in https://github.com/cockroachdb/cockroach/issues/43754

tbg commented:
CI failed because the created filenames could get too long. I made them more succinct (and generally brushed this PR up).

tbg commented:
bors r=andreimatei
